pkg/search/postgres/joins.go

```diff
@@ -185,5 +190,77 @@ func getJoinsAndFields(src *walker.Schema, q *v1.Query) ([]innerJoin, map[string
 		}
 	}
 
+	joinTreeRoot.removeUnnecessaryRelations(reachableFields)
+
 	return joinTreeRoot.toInnerJoins(), reachableFields
 }
+
+// removeUnnecessaryRelations removes inner join tables where the same column
+// is used by the previous and next table in the join chain. i.e.
+// a INNER JOIN b ON a.id = b.same_column
+// b INNER JOIN c ON b.same_column = c.id
+// If table b is not used in any other way, we can remove it from the join chain.
+// Outcome: a INNER JOIN c ON a.id = c.id
+func (j *joinTreeNode) removeUnnecessaryRelations(requiredFields map[string]searchFieldMetadata) {
+	if j == nil {
+		return
+	}
+
+	requiredTables := make(map[string]struct{})
+	for _, fieldMetadata := range requiredFields {
+		requiredTables[fieldMetadata.baseField.Schema.Table] = struct{}{}
+	}
+
+	rootChildren := make(map[*joinTreeNode][]walker.ColumnNamePair)
+	for child, columnPairs := range j.children {
+		child.removeUnnecessaryRelations(requiredFields)
+
+		if _, isRequired := requiredTables[child.currNode.Table]; isRequired {
+			rootChildren[child] = columnPairs
+
+			continue
+		}
+
+		childColumns := make(map[string]string)
+		for _, pair := range columnPairs {
+			childColumns[pair.ColumnNameInOtherSchema] = pair.ColumnNameInThisSchema
+		}
+
+		childChildren := make(map[*joinTreeNode][]walker.ColumnNamePair)
+		for childChild, childColumnPairs := range child.children {
+			if len(columnPairs) != len(childColumnPairs) {
+				childChildren[childChild] = childColumnPairs
+
+				continue
+			}
+
+			rootColumnPairs := make([]walker.ColumnNamePair, 0)
```

Comment:
microoptimisation
```suggestion
			rootColumnPairs := make([]walker.ColumnNamePair, len(childColumnPairs))
```

Reply:
This could break logic. Because I'm using `len(rootColumnPairs)` afterward to check that all columns are processed.
I'll change it to `0, len(childColumnPairs)` - because it will never grow bigger.